relengapi/blueprints/slaveloan/__init__.py

```diff
@@ -0,0 +1,233 @@
+    "Get the history associated with this loan"
+    # XXX: Use permissions to ensure admin | loanee
+    session = g.db.session('relengapi')
+    histories = session.query(History) \
+                       .filter(History.loan_id == loanid) \
+                       .order_by(asc(History.timestamp))
+    return [h.to_wsme() for h in histories.all()]
+
+
+@bp.route('/loans/all')
+@p.slaveloan.admin.require()
+@apimethod([rest.Loan])
+def get_all_loans():
+    "Get the list of all loans you can see"
+    # XXX: Use permissions to filter if not an admin
+    session = g.db.session('relengapi')
+    loans = session.query(Loans)
+    return [l.to_wsme() for l in loans.all()]
+
+
+@bp.route('/loans/new', methods=['POST'])
+@p.slaveloan.admin.require()
+@apimethod(None, body=rest.LoanAdminRequest)
+def new_loan_from_admin(body):
+    "Creates a new loan entry"
+    if not body.status:
+        raise BadRequest("Missing Status Field")
+    if not body.ldap_email:
+        raise BadRequest("Missing LDAP E-Mail")
+    if not body.bugzilla_email:
+        raise BadRequest("Missing Bugzilla E-Mail")
+    if body.status != 'PENDING':
+        if not body.fqdn:
+            raise BadRequest("Missing Machine FQDN")
+        if not body.ipaddr:
+            raise BadRequest("Missing Machine IP Address")
+
+    session = g.db.session('relengapi')
+    try:
+        if body.status != 'PENDING':
+            m = Machines.as_unique(session, fqdn=body.fqdn,
+                                   ipaddr=body.ipaddr)
+        h = Humans.as_unique(session, ldap=body.LDAP,
+                             bugzilla=body.bugzilla)
+    except sa.exc.IntegrityError:
+        raise InternalServerError("Integrity Error from Database, please retry.")
+
+    if body.status != 'PENDING':
+        l = Loans(status=body.status, human=h, machine=m)
+    else:
+        l = Loans(status=body.status, human=h)
+    history = History(for_loan=l,
+                      timestamp=tz.utcnow(),
+                      msg="Adding to slave loan tool via admin interface")
+    session.add(l)
+    session.add(history)
+    session.commit()
+    return None  # ?rest.WSME_New_Loan({'loan': l.to_wsme()})
+
+
+@bp.route('/loans/request', methods=['POST'])
+@apimethod(None, body=rest.LoanRequest)
+def new_loan_request(body):
+    "User Loan Requesting, returns the id of the loan"
+    if not body.ldap_email:
+        raise BadRequest("Missing LDAP E-Mail")
+    if not body.requested_slavetype:
+        raise BadRequest("Missing slavetype")
+
+    slavetype = slave_to_slavetype(body.requested_slavetype)
+    if not slavetype:
+        raise BadRequest("Unsupported slavetype")
+
+    if not body.bugzilla_email:
+        # Set bugzilla e-mail to ldap e-mail by default
+        body.bugzilla_email = body.ldap_email
+
+    session = g.db.session('relengapi')
+    try:
+        h = Humans.as_unique(session, ldap=body.ldap_email,
+                             bugzilla=body.bugzilla_email)
```

Comment:
What happens if I enter the same LDAP email but a different Bugzilla email?  Will this try to create a new row with the same LDAP and different Bugzilla (and fail because the LDAP column is unique)?  Or will it update the existing row?  Or just ignore the supplied Bugzilla email?